File: ibus-engine/tests/test_context_buffer.py
import pytest
import sys, pathlib
sys.path.insert(0, str(pathlib.Path(__file__).parent.parent))

from engine.context_buffer import ContextBuffer


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------


def buf(*chars: str) -> ContextBuffer:
    cb = ContextBuffer()
    for ch in chars:
        cb.append_char(ch)
    return cb


# ---------------------------------------------------------------------------
# context_text
# ---------------------------------------------------------------------------


def test_context_text_empty():
    assert ContextBuffer().context_text == ""


def test_context_text_after_typing():
    cb = buf("h", "e", "l", "l", "o")
    assert cb.context_text == "hello"


def test_context_text_after_accept_next_word():
    cb = buf("t", "y", "p", "e", " ")
    cb.set_suggestion("fast")
    cb.accept_next_word()
    assert cb.context_text == "type fast"


def test_context_text_after_accept_all():
    cb = buf("t", "y", "p", "e", " ")
    cb.set_suggestion("fast now")
    cb.accept_all()
    assert cb.context_text == "type fast now"


# ---------------------------------------------------------------------------
# append_char
# ---------------------------------------------------------------------------


def test_append_char_grows_buffer():
    cb = ContextBuffer()
    cb.append_char("a")
    cb.append_char("b")
    assert cb.context_text == "ab"


def test_append_char_dismisses_active_suggestion():
    cb = ContextBuffer()
    cb.set_suggestion("hello world")
    cb.append_char("x")
    assert not cb.has_suggestion
    assert cb.context_text == "x"


# ---------------------------------------------------------------------------
# backspace
# ---------------------------------------------------------------------------


def test_backspace_removes_last_char():
    cb = buf("a", "b", "c")
    cb.backspace()
    assert cb.context_text == "ab"


def test_backspace_on_empty_buffer_no_crash():
    cb = ContextBuffer()
    cb.backspace()  # must not raise
    assert cb.context_text == ""


def test_backspace_with_suggestion_dismisses_only():
    cb = buf("h", "i")
    cb.set_suggestion("there")
    cb.backspace()
    assert not cb.has_suggestion
    assert cb.context_text == "hi"  # char NOT removed


def test_backspace_without_suggestion_removes_char():
    cb = buf("h", "i")
    cb.backspace()
    assert cb.context_text == "h"


# ---------------------------------------------------------------------------
# set_suggestion / dismiss / has_suggestion
# ---------------------------------------------------------------------------


def test_set_suggestion_stores_text():
    cb = ContextBuffer()
    cb.set_suggestion("hello world")
    assert cb.suggestion == "hello world"
    assert cb.has_suggestion


def test_set_suggestion_empty_string_clears():
    cb = ContextBuffer()
    cb.set_suggestion("something")
    cb.set_suggestion("")
    assert not cb.has_suggestion
    assert cb.suggestion is None


def test_set_suggestion_replaces_previous():
    cb = ContextBuffer()
    cb.set_suggestion("first")
    cb.set_suggestion("second")
    assert cb.suggestion == "second"


def test_dismiss_clears_suggestion():
    cb = ContextBuffer()
    cb.set_suggestion("hello")
    cb.dismiss()
    assert not cb.has_suggestion


def test_dismiss_on_no_suggestion_no_crash():
    cb = ContextBuffer()
    cb.dismiss()  # must not raise
    assert not cb.has_suggestion


def test_dismiss_does_not_modify_buffer():
    cb = buf("a", "b")
    cb.set_suggestion("cd")
    cb.dismiss()
    assert cb.context_text == "ab"


# ---------------------------------------------------------------------------
# accept_next_word
# ---------------------------------------------------------------------------


def test_accept_next_word_single_word_no_space():
    cb = ContextBuffer()
    cb.set_suggestion("hello")
    word = cb.accept_next_word()
    assert word == "hello"
    assert not cb.has_suggestion


def test_accept_next_word_returns_word_with_trailing_space():
    cb = ContextBuffer()
    cb.set_suggestion("quick brown fox")
    word = cb.accept_next_word()
    assert word == "quick "


def test_accept_next_word_advances_through_suggestion():
    cb = ContextBuffer()
    cb.set_suggestion("one two three")
    assert cb.accept_next_word() == "one "
    assert cb.accept_next_word() == "two "
    assert cb.accept_next_word() == "three"
    assert not cb.has_suggestion


def test_accept_next_word_appends_to_buffer():
    cb = buf("s", "t", "a", "r", "t", " ")
    cb.set_suggestion("here now")
    cb.accept_next_word()
    assert cb.context_text == "start here "


def test_accept_next_word_exhausts_suggestion_on_last_word():
    cb = ContextBuffer()
    cb.set_suggestion("last")
    cb.accept_next_word()
    assert cb.suggestion is None
    assert not cb.has_suggestion


def test_accept_next_word_no_suggestion_returns_empty():
    cb = ContextBuffer()
    assert cb.accept_next_word() == ""


def test_accept_next_word_full_sequence_appended_to_buffer():
    cb = ContextBuffer()
    cb.set_suggestion("the lazy dog")
    cb.accept_next_word()
    cb.accept_next_word()
    cb.accept_next_word()
    assert cb.context_text == "the lazy dog"


# ---------------------------------------------------------------------------
# accept_all
# ---------------------------------------------------------------------------


def test_accept_all_returns_full_remaining_suggestion():
    cb = ContextBuffer()
    cb.set_suggestion("quick brown fox")
    text = cb.accept_all()
    assert text == "quick brown fox"


def test_accept_all_clears_suggestion():
    cb = ContextBuffer()
    cb.set_suggestion("something")
    cb.accept_all()
    assert not cb.has_suggestion


def test_accept_all_appends_to_buffer():
    cb = buf("a", " ")
    cb.set_suggestion("big idea")
    cb.accept_all()
    assert cb.context_text == "a big idea"


def test_accept_all_no_suggestion_returns_empty():
    cb = ContextBuffer()
    assert cb.accept_all() == ""


def test_accept_all_after_partial_accept():
    cb = ContextBuffer()
    cb.set_suggestion("one two three")
    cb.accept_next_word()           # consumes "one "
    rest = cb.accept_all()
    assert rest == "two three"
    assert cb.context_text == "one two three"
    assert not cb.has_suggestion


# ---------------------------------------------------------------------------
# Buffer overflow (maxlen=500)
# ---------------------------------------------------------------------------


def test_buffer_respects_maxlen():
    cb = ContextBuffer()
    for _ in range(600):
        cb.append_char("x")
    assert len(cb.context_text) == 500


def test_buffer_drops_oldest_chars():
    cb = ContextBuffer()
    for i in range(500):
        cb.append_char("a")
    cb.append_char("z")
    assert cb.context_text[-1] == "z"
    assert len(cb.context_text) == 500
    assert cb.context_text[0] == "a"


def test_accept_all_respects_maxlen():
    cb = ContextBuffer()
    for _ in range(490):
        cb.append_char("a")
    cb.set_suggestion("x" * 20)
    cb.accept_all()
    assert len(cb.context_text) == 500


# ---------------------------------------------------------------------------
# reset
# ---------------------------------------------------------------------------


def test_reset_clears_buffer():
    cb = buf("a", "b", "c")
    cb.reset()
    assert cb.context_text == ""


def test_reset_clears_suggestion():
    cb = ContextBuffer()
    cb.set_suggestion("hello world")
    cb.reset()
    assert not cb.has_suggestion
    assert cb.suggestion is None


def test_reset_clears_both():
    cb = buf("t", "y", "p", "e", " ")
    cb.set_suggestion("this")
    cb.reset()
    assert cb.context_text == ""
    assert not cb.has_suggestion


def test_reset_on_empty_buffer_no_crash():
    cb = ContextBuffer()
    cb.reset()  # must not raise
    assert cb.context_text == ""
    assert not cb.has_suggestion


def test_reset_allows_fresh_use():
    cb = buf("h", "e", "l", "l", "o")
    cb.set_suggestion("world")
    cb.reset()
    cb.append_char("x")
    assert cb.context_text == "x"
    assert not cb.has_suggestion

File: fcitx5-engine/tests/test_context_buffer.cpp
#include <catch2/catch_test_macros.hpp>
#include "context_buffer.h"

static ContextBuffer filled(const char* chars) {
    ContextBuffer cb;
    for (const char* p = chars; *p; ++p) cb.appendChar(*p);
    return cb;
}

// ── contextText ───────────────────────────────────────────────────────────────

TEST_CASE("contextText empty", "[context_text]") {
    REQUIRE(ContextBuffer().contextText() == "");
}

TEST_CASE("contextText after typing", "[context_text]") {
    REQUIRE(filled("hello").contextText() == "hello");
}

TEST_CASE("contextText after acceptNextWord", "[context_text]") {
    auto cb = filled("type ");
    cb.setSuggestion("fast");
    cb.acceptNextWord();
    REQUIRE(cb.contextText() == "type fast");
}

TEST_CASE("contextText after acceptAll", "[context_text]") {
    auto cb = filled("type ");
    cb.setSuggestion("fast now");
    cb.acceptAll();
    REQUIRE(cb.contextText() == "type fast now");
}

// ── appendChar ────────────────────────────────────────────────────────────────

TEST_CASE("appendChar grows buffer", "[append_char]") {
    ContextBuffer cb;
    cb.appendChar('a');
    cb.appendChar('b');
    REQUIRE(cb.contextText() == "ab");
}

TEST_CASE("appendChar dismisses active suggestion", "[append_char]") {
    ContextBuffer cb;
    cb.setSuggestion("hello world");
    cb.appendChar('x');
    REQUIRE_FALSE(cb.hasSuggestion());
    REQUIRE(cb.contextText() == "x");
}

// ── backspace ─────────────────────────────────────────────────────────────────

TEST_CASE("backspace removes last char", "[backspace]") {
    auto cb = filled("abc");
    cb.backspace();
    REQUIRE(cb.contextText() == "ab");
}

TEST_CASE("backspace on empty buffer no crash", "[backspace]") {
    ContextBuffer cb;
    cb.backspace();
    REQUIRE(cb.contextText() == "");
}

TEST_CASE("backspace with suggestion dismisses only", "[backspace]") {
    auto cb = filled("hi");
    cb.setSuggestion("there");
    cb.backspace();
    REQUIRE_FALSE(cb.hasSuggestion());
    REQUIRE(cb.contextText() == "hi");
}

TEST_CASE("backspace without suggestion removes char", "[backspace]") {
    auto cb = filled("hi");
    cb.backspace();
    REQUIRE(cb.contextText() == "h");
}

// ── setSuggestion / dismiss / hasSuggestion ───────────────────────────────────

TEST_CASE("setSuggestion stores text", "[suggestion]") {
    ContextBuffer cb;
    cb.setSuggestion("hello world");
    REQUIRE(cb.suggestion() != nullptr);
    REQUIRE(*cb.suggestion() == "hello world");
    REQUIRE(cb.hasSuggestion());
}

TEST_CASE("setSuggestion empty string clears", "[suggestion]") {
    ContextBuffer cb;
    cb.setSuggestion("something");
    cb.setSuggestion("");
    REQUIRE_FALSE(cb.hasSuggestion());
    REQUIRE(cb.suggestion() == nullptr);
}

TEST_CASE("setSuggestion replaces previous", "[suggestion]") {
    ContextBuffer cb;
    cb.setSuggestion("first");
    cb.setSuggestion("second");
    REQUIRE(*cb.suggestion() == "second");
}

TEST_CASE("dismiss clears suggestion", "[suggestion]") {
    ContextBuffer cb;
    cb.setSuggestion("hello");
    cb.dismiss();
    REQUIRE_FALSE(cb.hasSuggestion());
}

TEST_CASE("dismiss on no suggestion no crash", "[suggestion]") {
    ContextBuffer cb;
    cb.dismiss();
    REQUIRE_FALSE(cb.hasSuggestion());
}

TEST_CASE("dismiss does not modify buffer", "[suggestion]") {
    auto cb = filled("ab");
    cb.setSuggestion("cd");
    cb.dismiss();
    REQUIRE(cb.contextText() == "ab");
}

// ── acceptNextWord ────────────────────────────────────────────────────────────

TEST_CASE("acceptNextWord single word no space", "[accept_next_word]") {
    ContextBuffer cb;
    cb.setSuggestion("hello");
    REQUIRE(cb.acceptNextWord() == "hello");
    REQUIRE_FALSE(cb.hasSuggestion());
}

TEST_CASE("acceptNextWord returns word with trailing space", "[accept_next_word]") {
    ContextBuffer cb;
    cb.setSuggestion("quick brown fox");
    REQUIRE(cb.acceptNextWord() == "quick ");
}

TEST_CASE("acceptNextWord advances through suggestion", "[accept_next_word]") {
    ContextBuffer cb;
    cb.setSuggestion("one two three");
    REQUIRE(cb.acceptNextWord() == "one ");
    REQUIRE(cb.acceptNextWord() == "two ");
    REQUIRE(cb.acceptNextWord() == "three");
    REQUIRE_FALSE(cb.hasSuggestion());
}

TEST_CASE("acceptNextWord appends to buffer", "[accept_next_word]") {
    auto cb = filled("start ");
    cb.setSuggestion("here now");
    cb.acceptNextWord();
    REQUIRE(cb.contextText() == "start here ");
}

TEST_CASE("acceptNextWord exhausts suggestion on last word", "[accept_next_word]") {
    ContextBuffer cb;
    cb.setSuggestion("last");
    cb.acceptNextWord();
    REQUIRE(cb.suggestion() == nullptr);
    REQUIRE_FALSE(cb.hasSuggestion());
}

TEST_CASE("acceptNextWord no suggestion returns empty", "[accept_next_word]") {
    REQUIRE(ContextBuffer().acceptNextWord() == "");
}

TEST_CASE("acceptNextWord full sequence appended to buffer", "[accept_next_word]") {
    ContextBuffer cb;
    cb.setSuggestion("the lazy dog");
    cb.acceptNextWord();
    cb.acceptNextWord();
    cb.acceptNextWord();
    REQUIRE(cb.contextText() == "the lazy dog");
}

// ── acceptAll ─────────────────────────────────────────────────────────────────

TEST_CASE("acceptAll returns full remaining suggestion", "[accept_all]") {
    ContextBuffer cb;
    cb.setSuggestion("quick brown fox");
    REQUIRE(cb.acceptAll() == "quick brown fox");
}

TEST_CASE("acceptAll clears suggestion", "[accept_all]") {
    ContextBuffer cb;
    cb.setSuggestion("something");
    cb.acceptAll();
    REQUIRE_FALSE(cb.hasSuggestion());
}

TEST_CASE("acceptAll appends to buffer", "[accept_all]") {
    auto cb = filled("a ");
    cb.setSuggestion("big idea");
    cb.acceptAll();
    REQUIRE(cb.contextText() == "a big idea");
}

TEST_CASE("acceptAll no suggestion returns empty", "[accept_all]") {
    REQUIRE(ContextBuffer().acceptAll() == "");
}

TEST_CASE("acceptAll after partial accept", "[accept_all]") {
    ContextBuffer cb;
    cb.setSuggestion("one two three");
    cb.acceptNextWord();  // consumes "one "
    REQUIRE(cb.acceptAll() == "two three");
    REQUIRE(cb.contextText() == "one two three");
    REQUIRE_FALSE(cb.hasSuggestion());
}

// ── Buffer overflow (maxlen=500) ──────────────────────────────────────────────

TEST_CASE("buffer respects maxlen", "[overflow]") {
    ContextBuffer cb;
    for (int i = 0; i < 600; ++i) cb.appendChar('x');
    REQUIRE(cb.contextText().size() == 500u);
}

TEST_CASE("buffer drops oldest chars", "[overflow]") {
    ContextBuffer cb;
    for (int i = 0; i < 500; ++i) cb.appendChar('a');
    cb.appendChar('z');
    auto t = cb.contextText();
    REQUIRE(t.size() == 500u);
    REQUIRE(t.back() == 'z');
    REQUIRE(t.front() == 'a');
}

TEST_CASE("acceptAll respects maxlen", "[overflow]") {
    ContextBuffer cb;
    for (int i = 0; i < 490; ++i) cb.appendChar('a');
    cb.setSuggestion(std::string(20, 'x'));
    cb.acceptAll();
    REQUIRE(cb.contextText().size() == 500u);
}

// ── reset ─────────────────────────────────────────────────────────────────────

TEST_CASE("reset clears buffer", "[reset]") {
    auto cb = filled("abc");
    cb.reset();
    REQUIRE(cb.contextText() == "");
}

TEST_CASE("reset clears suggestion", "[reset]") {
    ContextBuffer cb;
    cb.setSuggestion("hello world");
    cb.reset();
    REQUIRE_FALSE(cb.hasSuggestion());
    REQUIRE(cb.suggestion() == nullptr);
}

TEST_CASE("reset clears both", "[reset]") {
    auto cb = filled("type ");
    cb.setSuggestion("this");
    cb.reset();
    REQUIRE(cb.contextText() == "");
    REQUIRE_FALSE(cb.hasSuggestion());
}

TEST_CASE("reset on empty buffer no crash", "[reset]") {
    ContextBuffer cb;
    cb.reset();
    REQUIRE(cb.contextText() == "");
    REQUIRE_FALSE(cb.hasSuggestion());
}

TEST_CASE("reset allows fresh use", "[reset]") {
    auto cb = filled("hello");
    cb.setSuggestion("world");
    cb.reset();
    cb.appendChar('x');
    REQUIRE(cb.contextText() == "x");
    REQUIRE_FALSE(cb.hasSuggestion());
}
